cpp/src/arrow/extension/tensor_array.cc

```diff
@@ -0,0 +1,161 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/extension/tensor_array.h"
+#include "arrow/extension_type.h"
+
+#include "arrow/json/rapidjson_defs.h"  // IWYU pragma: keep
+#include "arrow/tensor.h"
+#include "arrow/util/logging.h"
+
+#include <rapidjson/rapidjson.h>
+#include <rapidjson/document.h>
+#include <rapidjson/stringbuffer.h>
+#include <rapidjson/writer.h>
+
+namespace rj = arrow::rapidjson;
+
+namespace arrow {
+namespace extension {
+
+std::string TensorArrayType::extension_name() const { return "arrow.fixed_size_tensor"; }
+
+size_t TensorArrayType::ndim() const { return shape_.size(); }
+
+std::vector<int64_t> TensorArrayType::shape() const { return shape_; }
+
+std::vector<int64_t> TensorArrayType::strides() const {
+  std::vector<int64_t> strides;
+  const auto& element_type =
+      internal::checked_cast<const FixedWidthType&>(*storage_type());
+
+  if (is_row_major_) {
+    DCHECK_OK(internal::ComputeRowMajorStrides(element_type, shape_, &strides));
+  } else {
+    DCHECK_OK(internal::ComputeColumnMajorStrides(element_type, shape_, &strides));
+  }
+
+  return strides;
+}
+
+std::shared_ptr<DataType> TensorArrayType::storage_type() const { return storage_type_; }
+
+std::string TensorArrayType::order() const { return is_row_major_ ? "C" : "F"; }
+
+bool TensorArrayType::ExtensionEquals(const ExtensionType& other) const {
+  if (extension_name() != other.extension_name()) {
+    return false;
+  }
+
+  const auto& other_ext = static_cast<const TensorArrayType&>(other);
+  bool equals = storage_type()->Equals(other_ext.storage_type());
+  equals &= shape() == other_ext.shape();
+  equals &= strides() == other_ext.strides();
+  return equals;
+}
+
+std::string TensorArrayType::Serialize() const {
+  rj::Document document;
+  document.SetObject();
+  rj::Document::AllocatorType& allocator = document.GetAllocator();
+
+  rj::Value shape(rj::kArrayType);
+  for (auto v : shape_) {
+    shape.PushBack(v, allocator);
+  }
+  document.AddMember(rj::Value("shape", allocator), shape, allocator);
+  document.AddMember(rj::Value("order", allocator),
+                     rj::Value(is_row_major_ ? "C" : "F", allocator), allocator);
+  if (!document.HasMember("metadata")) {
+    document.AddMember(rj::Value("metadata", allocator), rj::Value(metadata_, allocator),
+                       allocator);
+  }
+
+  rj::StringBuffer buffer;
+  rj::Writer<rj::StringBuffer> writer(buffer);
+  document.Accept(writer);
+  return buffer.GetString();
+}
+
+Result<std::shared_ptr<DataType>> TensorArrayType::Deserialize(
+    std::shared_ptr<DataType> storage_type, const std::string& serialized_data) const {
+  rj::Document document;
+  document.Parse(serialized_data);
```

Comment:
Do you have an idea how hard it would be to parse this without rapid-json? (it's very limited json, so it might be doable) Just wondering it would be worth having this available without requiring this dependency.

Reply:
rapidjson is a very light-weight dependency that we can bundle at build time, so I don't think it would be worth the hassle.

Reply:
Is this what you mean? https://github.com/apache/arrow/pull/8510/files#diff-768505ba2d27dc99edc6880c892abd7a9383196bba6131981a8c0420b7c3c0bcR54

Reply:
The C++ minimal build example is still failing, so I suppose that is not sufficient. 

I think you need to update the CMakeLists.txt file where you added tensor_array.cc to do this conditionally, like is done for example here:

https://github.com/apache/arrow/blob/5765aa29b5ca93dae1a5fe8c7efce05dcd9f73c8/cpp/src/arrow/CMakeLists.txt#L520-L522